Commit vs rollbacks

pgrange opened the pull request:
Fixes #784

☕ Introduces `rollbackAndForward` on mock chain to simulate rollbacks during model spec executions.

☕ Adds ADR23 to provide a rationale for the following changes.

☕ Removes the high-level handling of rollbacks in `HeadLogic`, but keeps a `LocalChainState` in `Chain.Direct` which is rolled back and kept up-to-date to do consistent chain following. This gets rid of the problematic race condition between `setChainState` and `chainCallback`. The `chainState` in `headState` is kept for persistency reasons.

☕ Adds tests to `HandlersSpec` which do cover the newly moved-into-chain-layer `rollback` logic.

☕ Drops the `RolledBack` server output as clients cannot do anything about it.

☕ Small refactorings across `MockChain`, `BehaviorSpec`, `Model` modules.

---

<!-- Tick off or strike-through / remove if not applicable -->
* [x] CHANGELOG updated
* [x] Documentation updated
* [x] Added and/or updated haddocks
* [x] No new TODOs introduced or explained herafter

ffakenz committed:
Split chain callback

The chaincallback currently takes care of everything:
* updating the chainState inside the headState
* putting new events to the queue

We want to split that to prepare for the change where we only
put event to the queue to solve some synchronization problems.

ffakenz committed:
Add property to check: head opens if all participants commit

ffakenz committed:
Prepare simulaton of rollback and forward on every tick

When introducing rollbacks in the mockchain we can see the code
is not ready to support it. We inhibit this test in the pushed
code so we keep the C.I. green until the code can deal with that.

pgrange committed:
Managing the chainState in DirectChain

pgrange committed:
Link threads so that we get exception in case of failure

pgrange committed:
FIX wait forever for head is initialized

If we need the head to be initialized, we should really wait
forever for that to happen.

Previously, we could be in a situation where the head is
never initialized but we waited _long enough_ and _decide_
to commit anyway, ending up in strange and confusing errors.

pgrange committed:
Make it obvious that chainComponent is a different function

pgrange committed:
Wait for HeadIsInitialing/HeadIsAborted when performing Init/Abort actions

ch1bo committed:
Remove now unused waitForHeadIsInitializing

ch1bo committed:
Enable intermittent rollbacks in MockChain

This only passes the Model based tests as we disable arbitrary
generation of NewTx payments in an open head. The rollbacks would trip
the model when the hydra-node purges its state when rolling back "past
open".

However, we can still have basic conflict-free liveness tested and
exercise the rollback handling code in between Init/Commit/Open states.

pgrange committed:
Refactor waitForOpen

pgrange committed:
Problems with rollback management

We do not handle rollbacks correctly:
* a client would not now what to do with a rollback event
* we do not handle these events correctly

[#185](https://github.com/input-output-hk/hydra/issues/185) is about
dealing appropriately with the rollback.

With [#185](https://github.com/input-output-hk/hydra/issues/185)
we will be able to really solve the issues in an appropriate way for
both, the client using a hydra-node and the internals of our system.
In the mean time, we remove the rollback logic from Hydra head so that:
* we do not send rollback events to the client
* we _bet_ that the seen transactions will be seen again in the new fork
  of the chain

TODO:
* add the chainState to the rollback event so that we remove all the rollback
  logic from HeadLogic

pgrange commented:
Commit [Problems with rollback managemen](https://github.com/input-output-hk/hydra/pull/827/commits/d0b8df697ed65bf07a060e0b40787604992a7003) mainly remove a lot of rollback logic from the head logic. It has an impact on what's been done in #184 

This were the requirements from #184:


## ✅ The hydra-node does not crash when a rollback occurs 

This is still ensured by the "rolling back & forward" tests from BehaviorSpec:
We make the chain roll back and forward and then check that we can keep using the node without problem.

It is also ensured by the ModelSpec and the fact that the MockChain now rollbacks and the tests pass.

## ❌ A ServerOutput is emitted to the client when a rollback occurs 

We discovered that the rollback server output is of no use as it is right now. As a client, we get the message _rollback_ but, really, we have no idea what to do with that.

We could easily keep that message if we want but we decided to remove it for now as it's a sort of noise the client wouldn't no how to use. [#185 ](https://github.com/input-output-hk/hydra/issues/185)

## ✅ When a not-yet-open head is rolled back, clients can continue opening (i.e. commit, close or abort) the head after the rollback occurred (safety)

The property _check head opens if all participants commit_ in `ModelSpec` exercises this and proves it's possible for a head to be open in presence of rollbacks.

Although note that we only exercise rollbacks and forward were all the seen transactions will happen again in the fork, we do not consider situations were a transaction will _disappear_ from the fork.

Also, the chainState being rolled back, I could imagine a situation were the head logic would post a collecCom transaction when, from the chainState point of view, not everybody has committed yet. But for some reasons I was not able to reproduce this in the test... maybe I'm missing something here but even if it happens in real life, it should not introduce any security risk for the user. Worst case: a node crashing when trying to open the head?

## ✅ When an already-open head is rolled back "crossing opening point", it is acceptable that the off-chain communication state is reset and we could continue or contest (no liveliness)

I don't see how this requirement would have been impacted by this change.

ch1bo commented:
@pgrange Interesting proposal of not doing rollbacks at all on the HeadState!

> The hydra-node does not crash when a rollback occurs

This is not tested by the BehaviorSpec as it only covers the business logic and nothing on the chain layer. The ModelSpec with the MockChain including about 80% of the Direct chain should give higher confidence here.

> When a not-yet-open head is rolled back, clients can continue opening [...] Also, the chainState being rolled back, I could imagine a situation were the head logic would post a collecCom transaction when, from the chainState point of view, not everybody has committed yet.

These are the interesting cases and this can certainly happen. I do not think the node will crash though and expect more an error when posting a tx, but this is also happening today when everyone tries to collect after the last commit (which is the behavior on `master`).

Example scenario:
* Two party head
* Rollforward, Init tx -> observed, chain state: Initializing [] -> head state: Initializing []
* A submits commit tx
* Rollforward, Commit tx A -> observed, chain state: Initializing [A committed] -> head state: Initializing [A committed]
* B submits a commit tx
* Rollback -> chain state: Initializing []
  - Important: Head state **unchanged**   
* Rollforward, Commit tx B -> observed, chain state: Initializing [B committed] -> head state: Initializing [A committed, B committed]
  - node submits a collect tx
  - submission error because the chain layer cannot construct a collect tx from current state (or the resulting tx is just not valid)
* Rollforward, Commit tx A -> observed, chain state: Initializing [B committed, A committed] -> head state: Initializing [A committed, B committed]
  - Important: logic needs to be able to handle `Commit A` even though it already thinks A committed

> When an already-open head is rolled back "crossing opening point", it is acceptable that the off-chain communication state is reset and we could continue or contest (no liveliness)

This will be still true, depending in how we will implement this.
* Before, the logic rolled back the `HeadState` via the `previousRecoverableState` to the fork point.
* Without rollback handling in the logic layer, it depends in how we deal with a `OnCommitTx` observation when we are in `Open` head state.
  - A natural way to do interpret this is "The chain is always right", so we would reset the the head state to be `Initializing`.
    - In this case it is unclear what exactly is in this `Initializing` state? Shall we try to submit a collect tx right away?
  - Alternatively we could "Hope all head transactions are on the chain" and just do nothing (which does not drive the life cycle forward)

pgrange committed:
Rollback carries the chainState

Simplifies the HeadLogic: no need to look for the previous
head state just to find the chain state. The chain component
nows about the chain state to rollback to so it just sends
it in the event.

We don't use the ChainSlot in the event so we could remove it.
However I think it makes sense to share this information in the
event so, not sure, but I'll keep it for now with some sort of
separation of concern idea in mind: the chain component emits
the event that the chain has been rolled back to that point and
it's now in this state. The fact that the information is used
by the receiver or not is not is business.

pgrange committed:
Removing Rollback from API

pgrange committed:
Updating our ADRs about rollbacks

abailly committed:
Missing removal of RolledBack handler

ch1bo commented on docs/adr/2023-04-26_022-single-state.md:
```diff
@@ -0,0 +1,46 @@
+---
+slug: 22
+title: | 
+  22. Single state in Hydra.Node.
```
22 is already taken https://hydra.family/head-protocol/adr/22/

pgrange replied:
@ch1bo fixed

ch1bo commented on docs/core-concepts/rollbacks/index.md:
```diff
@@ -41,9 +41,4 @@ The consequences of a rollback on the Head's state are different depending at wh
-![](rollbacks-4.jpg)
-
-* When a relevant transaction is observed, eg. `OnInit` or `OnCommit` in the case depicted above, it induces a new `HeadState` which is linked to the _previous_ state,
-* When a rollback is observed on-chain, the `HeadLogic` is notified with `Rollback` event telling the number of steps that need to be rolled-back: The `HeadLogic` layer does not need to care about the details of what a `Point Block` is on-chain, it only stores a sequence of state that can be easily "rolled back" to a previous state.
```
This was even outdated :o

ch1bo commented on hydra-node/src/Hydra/Chain.hs:
```diff
@@ -204,7 +204,10 @@ data ChainEvent tx
-  | Rollback ChainSlot
+  | Rollback
+      { slot :: ChainSlot
+      , newChainState :: ChainStateType tx
```
Should: Only include `newChainState` and use the `IsChainState` type class which practically all `ChainStateType tx` instance need to be part of to determine `chainStateSlot newChainState`.

ch1bo replied:
Maybe even call it `rolledBackChainState` if you agree it would be even clearer.

ch1bo replied:
Done in https://github.com/input-output-hk/hydra/commit/76b341d26d7af357d7cdbf9bebefda38b362d852

ch1bo commented on hydra-node/src/Hydra/Chain.hs:
```diff
@@ -225,7 +228,7 @@ instance
 
 -- | A callback indicating receival of a potential Hydra transaction which is Maybe
 -- observing a relevant 'ChainEvent tx'.
```
Must: update haddock

ch1bo replied:
https://github.com/input-output-hk/hydra/commit/b0a830b03f3a877f43a2be4ccb327d89e282e315

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -187,7 +199,30 @@ chainSyncHandler tracer callback getTimeHandle ctx =
+    -- rollback the tvar to the chainSlotFromPoint
+    chainStates <- readTVarIO chainStateTVar
+    let (chainState, chainStates') = rollback point chainStates
+    atomically $ writeTVar chainStateTVar chainStates'
```
Must: use a single `atomically` to do the update.

This is to prevent race conditions in case the tvar would be used by multiple threads.
Also consider making rollback run in STM to avoid the tuple, i.e. `rollback :: SlotNo -> STM ChainStateAt` and call run it by `rolledBackChainState <- rollback (chainSlotFromPoint point)` .. that way the comment above also becomes redundant.

ch1bo replied:
Done in https://github.com/input-output-hk/hydra/commit/b533c8d260e24b236833d3b08c6f9ade69594184

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -202,10 +199,32 @@ instance (IsTx tx, Arbitrary (ChainStateType tx)) => Arbitrary (InitialState tx)
+--              -> Committed tx
+--              -> ChainStateType tx
+--              -> HeadId
+--              -> InitialState tx
```
Must: remove commented code

pgrange committed:
FIX ADR number

ch1bo commented on CHANGELOG.md:
```diff
@@ -10,6 +10,9 @@ changes.
 
 ## [0.10.0] - UNRELEASED
 
+- **BREAKING** Rollback has been removed from the API
```
Should: be precise here. I believe it was the `RolledBack` message that was removed?

I'm not 100% how we should structure these changes. This one here was changing the semantics of the internal working of hydra-node. not so much user related other than logs and API as you rightfully state here.

ch1bo commented on docs/adr/2022-04-13_018-single-state.md:
```diff
@@ -3,12 +3,12 @@ slug: 18
 ## Status
 
-Accepted
+Superseded by ADR 22
```
```suggestion
Superseded by [ADR 23](/adr/23)
```

ch1bo commented on docs/adr/2023-04-26_023-single-state.md:
```diff
@@ -0,0 +1,50 @@
+* published an event about this observation
+* update the headState with some `chainState 1`
+
+# The Node processes the event and emit a new headState with a previousRecoverableState in case a rollback later happens
```
These are headlines, you might want to use `1.` for numbered lists

```suggestion
1. The DirectChain component fetch some `chainState 0` from the `headState`

1. The DirectChain component observes a transaction and it

* publishes an event about this observation
* updates the `headState` with some `chainState 1`

1. The Node processes the event and emits a new `headState` with a `previousRecoverableState` in case a rollback later happens
```

ch1bo commented on docs/adr/2023-04-26_023-single-state.md:
```diff
@@ -0,0 +1,50 @@
+* We supersede [ADR 18](2022-04-13_018-single-state.md) with the current ADR.
+* A local chain state is re-introduced in the chain component, not shared with the head logic.
+* A copy of the chain state is kept in the head state to keep the benefits of [ADR 18](2022-04-13_018-single-state.md) regarding persistency.
+* The rollback event is removed from the API until [#185](https://github.com/input-output-hk/hydra/issues/185).
```
You mean the `RolledBack` `ServerOutput` is removed, right?

ch1bo commented on hydra-node/src/Hydra/Chain/Direct.hs:
```diff
@@ -182,11 +183,14 @@ withDirectChain ::
   Tracer IO DirectChainLog ->
   ChainConfig ->
   ChainContext ->
-  -- | Last known point on chain as loaded from persistence.
```
Should: move this to the `ChainStateAt` argument now.

IMO this haddock documentation was a keeper

ch1bo replied:
When reviving this I realized.. we need to not only load the latest chainState, but all the known chain states now! We cannot just initialize the direct chain with the latest / last known chain state as it is not linking to previous ones.

Scenario:
* Observe the head opening through several transactions
* Restart the node
* See a rollback before one of the Head transactions -> boom!
  - We would not be able to rollback and re-observe the following roll-forward of head transactions

pgrange replied:
Absolutely. If new want it to be coherent, the _chainState_ should point to the previous _chainState_ and that shall be stored to disk. Close to what we had in the HeadState I would say.

ch1bo replied:
Implemented in https://github.com/input-output-hk/hydra/commit/07dfbf1f533b0bf03a6ec4a0543c4ea5929507ed and 2-3 commits before that.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct.hs:
```diff
@@ -182,11 +183,14 @@ withDirectChain ::
+withDirectChain tracer config ctx wallet chainStateAt callback action = do
+  -- Last known point on chain as loaded from persistence.
+  let persistedPoint = recordedAt chainStateAt
+  -- TODO we should use NonEmptyList here instead
```
Yes, do it.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -176,9 +186,11 @@ chainSyncHandler ::
   -- | Contextual information about our chain connection.
   ChainContext ->
+  -- TVar containing the local ChainState
+  TVar m [ChainStateAt] ->
```
Should: Make the comment a haddock and maybe improve by rather saying that the latest chain state is in the front?

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -187,7 +199,30 @@ chainSyncHandler tracer callback getTimeHandle ctx =
+    atomically $ writeTVar chainStateTVar chainStates'
+    callback (Rollback (chainSlotFromPoint point) chainState)
+   where
+    rollback rollbackChainPoint chainStates =
```
Should: try to avoid nested `where` blocks

This is easily possible here as we do not use the lexical scope of `onRollBackward` anyways here.

ch1bo commented on hydra-node/src/Hydra/Node.hs:
```diff
@@ -262,23 +260,7 @@ createNodeState initialState = do
-    Nothing -> pure ()
-    Just chainEvent -> putEvent eq $ OnChainEvent{chainEvent}
+chainCallback eventQueue chainEvent = do
+  putEvent eventQueue $ OnChainEvent{chainEvent}
```
Could: move this back to the `Main` function

Not sure, but that could discourage adding any more logic to this (again).

ch1bo replied:
Done

ch1bo commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -464,8 +464,9 @@ spec = parallel $ do
-            waitUntil [n1] $ HeadIsInitializing testHeadId (fromList [alice])
+            -- We expect the node to still work and let us commit
+            send n1 (Commit (utxoRef 1))
+            waitUntil [n1] $ Committed testHeadId alice (utxoRef 1)
```
Was the previous assertion not working anymore?

pgrange replied:
The previous assertion could not work anymore.

Obviously we can not wait for a `Rolledback` event since it is not sent anymore but not only that, since there is no notion of rollback from the client perspective, it would not make any sense to expect to see the `HeadIsInitializing` again.

So, instead, we're just checking that, after a rollback, the head is still functional by committing stuff in it and observing the commit without any error.

pgrange replied:
/cc @ch1bo

ch1bo commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -479,9 +480,9 @@ spec = parallel $ do
-            waitUntil [n1] RolledBack
-            waitUntil [n1] $ Committed testHeadId alice (utxoRef 1)
-            waitUntil [n1] $ HeadIsOpen{headId = testHeadId, utxo = utxoRefs [1]}
+            -- We expect the node to still work and let us post L2 transactions
```
This means that we also expect that `HeadIsOpen` is **NOT** emmitted again, right?

pgrange replied:
I would rather say that we don't really care, in this test, if `HeadIsOpen` is emitted again. What we want to assert is that, once we've seen a head open, it's open and we can use it, no matter if there has been rollbacks.

But that's a good point because, looking at the HeadLogic, I don't like the smell of it. Eventhough the behavior would probably be fine, I think we can see `InvalidEvent` errors when the collectCom transaction is observed again but the HeadLogic consider the head as open already.

ch1bo replied:
We have the same taste of code then :)

Having `InvalidEvent` error outcomes now because we see a `OnCommit` or `OnCollectCom` when the head is already `Open`, is a smell indeed. I mentioned once that in this case we would need to change the head logic in a way that "the chain is always right" and we could identify already that "a rollback must have happened, we bet it all is fine though and stay tuned" is rather the outcome when seeing this event?

ch1bo commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -543,7 +544,8 @@ data TestHydraNode tx m = TestHydraNode
   { chainComponent :: HydraNode tx m -> m (HydraNode tx m)
   , tickThread :: Async m ()
-  , rollbackAndForward :: Natural -> m ()
+  , -- TODO remove the following if we really don't use it anymore
```
We do use it just above.

ch1bo commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -640,7 +642,7 @@ simulatedChainAndNetwork initialChainState = do
     -- Yield rollback events
     ns <- readTVarIO nodes
-    forM_ ns $ \n -> handleChainEvent n (Rollback $ chainStateSlot rolledBackChainState)
+    forM_ ns $ \n -> handleChainEvent n (Rollback (chainStateSlot rolledBackChainState) rolledBackChainState)
```
This is what I meant in another comment: We do not need to include the slot if we have the `chainState` already.

pgrange replied:
@ch1bo please see comment of commit [3f7735bbd5d](https://github.com/input-output-hk/hydra/pull/827/commits/3f7735bbd5d1e274b124a57260d618d143eee2b7) for the rationale behind this.

ch1bo replied:
Thanks for pointing me there. I'm not sure I buy the duplication vs. separation of concerns argument. Also, there is no need for the slot in the event anymore if we do not rollback in the head logic as we would only store the included state.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/HandlersSpec.hs:
```diff
@@ -50,6 +47,7 @@ import Hydra.Chain.Direct.State (
 import Hydra.Chain.Direct.TimeHandle (TimeHandle (slotToUTCTime), TimeHandleParams (..), genTimeParams, mkTimeHandle)
 import Hydra.Options (maximumNumberOfParties)
 import Test.Consensus.Cardano.Generators ()
+import Test.Hydra.Prelude
```
Should: group implicit imports above explicit imports

https://github.com/input-output-hk/hydra/wiki/Coding-Standards#proposal-use-of-imports

(should make it a non-proposal, but accepted :) )

pgrange replied:
@ch1bo I'm expecting the code formatter to deal with that kind of detail. Does your formatter arrange things differently for this file?

ch1bo replied:
HLS does add imports to the bottom of the list and the formatter will put it in alphabetical order AFAIK. There is no option about grouping implicit vs. explicit imports - that is a thing we have been doing, but not encoded in a tool.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/HandlersSpec.hs:
```diff
@@ -117,55 +117,48 @@ spec = do
     -- Generate a state and related transaction and a block containing it
     (ctx, st, tx, transition) <- pick genChainStateWithTx
-    let chainState = ChainStateAt{chainState = st, recordedAt = Nothing}
+    -- let chainState = ChainStateAt{chainState = st, recordedAt = Nothing}
```
Must: remove commented code

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/HandlersSpec.hs:
```diff
@@ -181,18 +174,16 @@ spec = do
+recordEventsHandler ctx _cs getTimeHandle = do
   eventsVar <- newTVarIO []
-  let handler = chainSyncHandler nullTracer (recordEvents eventsVar) getTimeHandle ctx
+  chainStateVar <- newTVarIO [_cs]
```
Must: not prefix used variables with `_`

ch1bo commented on hydra-node/test/Hydra/Model/MockChain.hs:
```diff
@@ -69,82 +75,113 @@ mockChainAndNetwork ::
-  return (ConnectToChain{..}, tickThread)
+  return
+    ( ConnectToChain
+        { rollbackAndForward = error "Not implemented, should never be called"
```
I think it would be nice if we would actually use this function to do the rollback + forward logic. We would still trigger it every now and then from the `tickThread`, but it would make it very clear that we could also deliberately call the `rollbackAndForward`?

ch1bo replied:
done in https://github.com/input-output-hk/hydra/commit/61939865c2f72a3773befb5d2aebc30bd77661ed

ch1bo commented on hydra-node/test/Hydra/Model/MockChain.hs:
```diff
@@ -69,82 +75,113 @@ mockChainAndNetwork ::
+          node
+            { hn = createMockNetwork node nodes
+            , oc = chainHandle
+            , nodeState
```
Should: not override `nodeState` as this is not relevant for mock chain semantics (anymore)

ch1bo commented on hydra-node/test/Hydra/ModelSpec.hs:
```diff
@@ -1,6 +1,7 @@
 {-# LANGUAGE DuplicateRecordFields #-}
 {-# LANGUAGE PatternSynonyms #-}
 {-# LANGUAGE ScopedTypeVariables #-}
+{-# OPTIONS_GHC -Wno-unused-do-bind #-}
```
Should: not need to disable this warning.

pgrange replied:
We do need to disable this warning with the code as it is now due to a quick check dynamic evolution and some outcome we don't use. I'll fix the code to remove the option.

ch1bo commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -87,12 +86,12 @@ import qualified Prelude
+    hydraParties :: [(SigningKey HydraKey, CardanoSigningKey)]
+  , -- | Expected consensus state
+    -- All nodes should be in the same state.
+    hydraState :: GlobalState
```
Should: use the same formatter version + config as everyone else.

I'm open to update further to a different version, but we need to be using the same :)

pgrange replied:
@ch1bo using fouromlu 0.9.0 now would mess with the whole file formatting in this branch. Is there a specific format issue I can solve here in the meantime?

ch1bo commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -455,21 +477,32 @@ instance
         forM_ nodes $ \(_, node) -> do
           lift (waitForUTxOToSpend mempty (to tx) (value tx) node) >>= \case
-            Left u -> throwIO $ TransactionNotObserved tx u
+            Left u -> error $ "Did not observe transaction " <> show tx <> " applied: " <> show u
```
Should: use the `throwIO` with exception if it is available

It should behave better than the `error` call (which is not properly forced in non-IO)

ch1bo commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -455,21 +477,32 @@ instance
+          outputs <- lift $ serverOutputs node
+          case find headIsOpen outputs of
+            Just _ -> pure ()
+            Nothing -> error $ "The head is not open for node " -- <> show node
```
Must: remove commented code (or do show the node)

ch1bo commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -517,9 +550,8 @@ performCommit ::
   nodes <- gets nodes
   case Map.lookup party nodes of
-    Nothing -> throwIO $ UnexpectedParty party
+    Nothing -> error $ "unexpected party " <> Hydra.Prelude.show party
```
Should: use `throwIO` instead of `error`

ch1bo commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -583,25 +610,60 @@ performNewTx party tx = do
+  isOpen _ (HeadIsOpen{} : rest) = isOpen True rest
+  -- Remember, if we introduce rollback again, we'll have to
+  -- take them into account here
+  -- isOpen _ (RolledBack{} : rest) = isOpen False rest
```
This comment and the haddock above are outdated and likely not be here anymore when we re-introduce the rollback handling or not applicable because we do things different.

Could: simplify this function again to just look for a HeadIsOpen, keeping the search from the back (how are `serverOutputs` ordered?) and put a NOTE that we should consider rollbacks once they are visible to clients again.

ch1bo replied:
done in https://github.com/input-output-hk/hydra/commit/45ccd47bb52ae4174a9da3f45019896065261962

ch1bo reviewed (changes requested):
This is taking shape, but many TODOs and things to clean up still. Also:

- [x] As we moved the rollback logic from the head logic to the chain layer, I am missing tests covering that the rollback happens correctly (in the list of chain states). Not sure exactly where though.

- [x] Load full sequence of chain states instead of only latest.

- [x] The MockChain should use the `rollbackAndForward` function to implement a rollback .. and forward semantics.

pgrange committed:
Update docs/adr/2022-04-13_018-single-state.md

Co-authored-by: Sebastian Nagel <ch1bo@users.noreply.github.com>

pgrange committed:
Useless TODO

pgrange committed:
Typo: not prefix used variables with _

pgrange committed:
use the throwIO with exception

It should behave better than the error call (which is not properly forced in non-IO)

v0d1ch committed:
Use NonEmpty list for ChainState

v0d1ch committed:
Use single atomically block in `chainSyncHandler`

v0d1ch committed:
Do not nest where blocks

v0d1ch committed:
Remove commented out code

v0d1ch committed:
Remove unwanted pragma

v0d1ch committed:
Throw instead of using error when we encounter `UnexpectedParty`

ch1bo committed:
FIX rolled back chainState in HandlersSpec

We currently have the slot and the chain state in the rollback event. If
we only use the chain slot and never use the included chainState, the
computation for rolling it back will never be forced and practically the
'rollback' function is not tested that way.

ch1bo committed:
Extract a local chain state handle

This hides the fact that internally a list of 'ChainStateAt' values is
kept and narrows down the interface of getting the latest chain state,
pushing a new one and rolling back to a previous one.

ch1bo committed:
Update docs/adr/2023-04-26_023-single-state.md

ch1bo committed:
Reformat HeadLogic

ch1bo committed:
Remove slot from Rollback event

This is redundant to the already included ChainStateType as every such
chain state needs have a ChainSlot via the IsChainState type class.

ch1bo committed:
Update/restore haddocks

ffakenz commented on docs/adr/2023-04-26_023-single-state.md:
```diff
@@ -0,0 +1,50 @@
+
+## Consequences
+
+* The rollback logic is removed from the HeadLogic and only maintained in the chain component.
```
maybe worth to mention why the HeadLogic can still work without to rollback logic.

ffakenz commented on hydra-node/test/Hydra/Model/MockChain.hs:
```diff
@@ -69,82 +85,113 @@ mockChainAndNetwork ::
+      Nothing ->
+        pure ()
+  sendRollBackward chain nbBlocks =
+    if False -- inhibiting the rollback until we can support it in the code
```
are we going to keep this flag?

ch1bo replied:
Good catch. The mock chain should either do random rollbacks or not. No dead code should remain.

ffakenz commented on hydra-node/test/Hydra/Model/MockChain.hs:
```diff
@@ -69,82 +85,113 @@ mockChainAndNetwork ::
+  chainComponent queue node = do
+    let Environment{party = ownParty, otherParties} = env node
+    let (vkey, vkeys) = findOwnCardanoKey ownParty seedKeys
+    let ctx =
```
i think this ctx is worth to be defined in the where clause.

ch1bo replied:
It uses `node` via `ownParty` so we cannot move it outside of the lexical hope.

ch1bo committed:
Rephrase changelog

ch1bo committed:
Re-indent, slightly reword and change title of ADR23

v0d1ch committed:
Make ChainStateAt recursive

- We want to be able to keep the history of chain states.

ch1bo committed:
Regenerate golden file because of changed ChainState

ch1bo committed:
Add test of LocalChainState to be able to resume

v0d1ch committed:
Introduce a rollBackward to the test

ch1bo committed:
Refactor genRollbackBlocks and fix assertion

We may see a rollback by 0 observable blocks, which is okay.

ch1bo committed:
Implement LocalChainState using the recursive ChainStateAt

This will correctly allow it to resume from the provided 'ChainStateAt'
even in presence of rollbacks.

ch1bo committed:
Remove dedicated chainCallback and refactor MockChain

ch1bo committed:
Reformat Model.hs

ch1bo committed:
Rename handles in BehaviorSpec and MockChain

ConnectToChain -> SimulatedChainNetwork
TestHydraNode -> TestHydraClient

ch1bo committed:
Push MockChain nodes TVar into mockChainAndNetwork

This will simplify any further refactoring of the mocked chain and
connected nodes are managed internally.

ch1bo committed:
Define rollbackAndForward for MockChain

Also refactor flushQueue and inline block creation.

ch1bo committed:
Fix documentation and simplify waitForOpen in Model

ch1bo committed:
Avoid realToFrac in a time conversion

ch1bo committed:
Resolve a FIXME

The comment was already saying that this is good practice and indeed in
the given context 'm' can be any IO-ish monad, so we keep the 'threads'
to clean them up in 'stopTheWorld'.

ch1bo reviewed (approved):
I continued this PR and think it's done now. Requires someone else's review.

ffakenz commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -131,6 +131,7 @@ class HasKnownUTxO a where
 data ChainStateAt = ChainStateAt
   { chainState :: ChainState
   , recordedAt :: Maybe ChainPoint
+  , previous :: Maybe ChainStateAt
```
I would like to "make invalid state unrepresentable" here.
I think we should refactor ChainStateAt into a sum type to avoid it.

```hs
data ChainStateAt 
  = ChainStateAtZero
  | ChainStateAt
    { chainState :: ChainState
    , recordedAt :: ChainPoint
    , previous :: ChainStateAt
    }
```

ch1bo replied:
I like that mantra, but what you suggest is not the same. We can have a `ChainStateAt` which has a `ChainState` that was `recordedAt` some point.. but still has no `previous` entry.

```
data ChainStateAt 
  = ChainStateAtZero
    { chainState :: ChainState
    , recordedAt :: ChainPoint
    }
  | ChainStateAt
    { chainState :: ChainState
    , recordedAt :: ChainPoint
    , previous :: ChainStateAt
    }
```

Is an alternative, but that is not clearer than the `Maybe` in `previous IMO.

pgrange committed:
Use actual cabal index for chap

That will remove the following warning when compiling:
```
Warning: Requested index-state 2023-03-22T09:20:07Z is newer than 'cardano-haskell-packages'! Falling back to older state (2023-03-21T14:44:35Z).
```
